### PR TITLE
Driver Assist

### PR DESCRIPTION
**Pull request by jbko6**
This needs to be tested on the bots. Also needs the substation locations as alignment points. Alignment points will also need to be tuned.

### CODE REVIEW

**Comment by jbko6** [comment]
Known issue: paths are being generated with heading, not rotation at the moment. As seen in https://github.com/mjansen4857/pathplanner/wiki/PathPlannerLib:-Java-Usage, the second parameter of `PathPoint` is heading, not rotation.

File: src/main/java/frc/team2412/robot/util/DriverAssist.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			double newDistance = robotTranslation.getDistance(alignmentPoses[i].getTranslation());
+			if (newDistance > distance) {
+				closestIndex = i;
+				distance = newDistance;
+			}
+		}
+		return alignmentPoses[closestIndex];
+	}
+
+	private static boolean isPoseTooFar(Pose2d robotPose, Pose2d alignmentPose) {
```
Comment (bobtseattle) [inline src/main/java/frc/team2412/robot/util/DriverAssist.java:54]: In addition to automation, driver assist could be helpful giving the driver feedback for how close they are to being aligned via red, green yellow feedback in shuffleboard  (or even feedback for the LEDs on the robot).

**Review by bobtseattle — COMMENTED** [review]
Some initial thoughts.

File: src/main/java/frc/team2412/robot/util/DriverAssist.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import frc.team2412.robot.subsystems.DrivebaseSubsystem;
+
+public class DriverAssist {
+	private static final PIDController ASSIST_TRANSLATION_PID = new PIDController(0, 0, 0);
+	private static final PIDController ASSIST_ROTATION_PID = new PIDController(0, 0, 0);
+	private static final PathConstraints ASSIST_CONSTRAINTS = new PathConstraints(1, 0.8);
+	private static final double MAX_ALIGNMENT_DISTANCE_METERS = 1.0;
+
+	// comments are relative to the driver, with numbers increasing from right to left
+	private static final Pose2d[] alignmentPoses = {
```
[the remark this answers is not in the document]
Comment (jbko6) [reply src/main/java/frc/team2412/robot/util/DriverAssist.java:22]: Yup, I didn't have time to add all the alignment points yet

**Comment by jbko6** [comment]
FYI for those working on this, this still needs to be able to mirror the path correctly when the blue alliance is used.

**Comment by KangarooKoala** [comment]
Oh, so it turns out PathPlanner doesn't mirror for us?

**Comment by jbko6** [comment]
It does, but then it mirrors your starting position too

**Review by TAKBS2412 — COMMENTED** [review]
I should point out that without accurate vision measurements the odometry may not be accurate after a while, or after driving over the charge station.

**Comment by jbko6** [comment]
Drafted until vision is in a good state

**Pull request closed without merging**